File: cpp_version/snake.cpp
//
//  main.cpp
//  line
//
//  Created by Nicolò Padovan on 07/11/2020.
//

#include <iostream>
#include <ncurses.h>
#include <time.h>
#include <array>
#include <deque>

using namespace std;
using Point = std::array<short, 2>;

class Snake {
    public:
        short w, h;
        std::deque<Point> history;
        unsigned int score = 0;
        array<short, 2> currentPosition = {10, 10};
        array<short, 2> foodPosition = {};
        unsigned char direction = 0; // 0 = UP, 1 = RIGHT, 2 = DOWN, 3 = LEFT
        time_t startTime = 0;
        short edges[2][2] = {{0,0},{0,0}};
        bool gameOver = FALSE;
        void addString(const char text[2], int y, int x);
        void start();
        void updateWindow();
        void updateDims();
        void reset();
        array<short int, 2> generateFood(bool ignore);
        void generateTail();
        Snake();
};

Snake::Snake() {
    curs_set(0);
    updateDims();

    if (Snake::w%2 != 0) {
        resizeterm(Snake::h, Snake::w-1);
        updateDims();
    }
    
    edges[0][1] = w;
    edges[1][1] = h;

    reset();
    start();
}

void Snake::reset() {
    startTime = time(0);
    gameOver = FALSE;
    score = 1;
    timeout(0);
    
    while (!history.empty()) {
        history.erase(history.begin());
    }
    currentPosition[0] = w/2; currentPosition[1] = h/2;
    if (currentPosition[0]%2 != 0) {
        currentPosition[0] += 1;
    }
    direction = 0;
    foodPosition[0] = 0; foodPosition[1] = 0;
    generateFood(TRUE);
}

void Snake::updateDims() {
    getmaxyx(stdscr, h, w);
}

void Snake::start() {
    while (TRUE) {
        while (!gameOver) {
            updateWindow();
        }

        timeout(-1);
//        clear();
//        wmove(stdscr, 0, 0);
//        wprintw(stdscr, "Game-over");
        getch();
        reset();
    }
}

void Snake::updateWindow() {
    time_t currentTime = difftime(time(0), startTime);
    
    if (currentTime > 0.1) {
        int key = getch();

        if (key == KEY_UP) {
            if (direction != 2) {
                direction = 0;
            }
        } else if (key == KEY_RIGHT) {
            if (direction != 3) {
                direction = 1;
            }
        } else if (key == KEY_DOWN) {
            if (direction != 0) {
                direction = 2;
            }
        } else if (key == KEY_LEFT) {
            if (direction != 1) {
                direction = 3;
            }
        }

        history.push_front(currentPosition);
        while (history.size() > score) {
            history.erase(history.end());
        }
        
        switch (direction) {
            case 0:
                currentPosition[1] -= 1;
                break;
            case 1:
                currentPosition[0] += 2;
                break;
            case 2:
                currentPosition[1] += 1;
                break;
            case 3:
                currentPosition[0] -= 2;
                break;
        }
        
        bool find = std::find(history.begin(), history.end(), currentPosition) != history.end();
        if (currentPosition[0] < edges[0][0] || currentPosition[0] > edges[0][1] || currentPosition[1] < edges[1][0] || currentPosition[1] > edges[1][1] || find) {
            gameOver = TRUE;
            return;
        } else {
            clear();
            array<short int, 2> foodPos = generateFood(FALSE);
            addString("FF", foodPos[1], foodPos[0]);
            addString("AA", currentPosition[1], currentPosition[0]);
            generateTail();
        }
        startTime = time(0);
    }
}

array<short int, 2> Snake::generateFood(bool ignore) {
    if (currentPosition == foodPosition || ignore) {
        score += 1;
        array<short int, 2> newFoodPos;
        while (TRUE) {
            short rand_1 = rand()%edges[0][1];
            short rand_2 = rand()%edges[1][1];

//            newFoodPos = {rand_1, rand_2};
            newFoodPos[0] = rand_1; newFoodPos[1] = rand_2;

            if (newFoodPos != currentPosition && newFoodPos[0]%2==0) {
                break;
            }
        }
        foodPosition = newFoodPos;
        return newFoodPos;
    } else {
        return foodPosition;
    }
}

void Snake::generateTail() {
    for(int i=0; i<history.size(); i++) {
        addString("AA", history[i][1], history[i][0]);
    }
}

void Snake::addString(const char text[2], int y, int x) {
    wmove(stdscr, y, x);
    addstr(text);
}

class Player {
public:
    array<short, 2> currentPosition;
    short direction = 0;
    short h, w;
    Player();
    void updateWindow();
};

Player::Player() {
    getmaxyx(stdscr, h, w);
    
}

void Player::updateWindow() {
    
}


int main() {
    
    initscr();
    keypad(stdscr, TRUE);
    start_color();
    curs_set(0);
    cbreak();
    noecho();
//    nodelay(stdscr, TRUE);
    
    Snake snake;
}

File: python_version/snake.py
import curses as cr
import logging
import os
from time import sleep, time
from numpy import random
import threading

class Snake:
    
    def __init__(self, screen):
        
        self.init_logger()
        cr.curs_set(0)
        self.updateDims(screen)
                
        screen.timeout(1)
                
        if self.w%2 != 0:
            cr.resizeterm(self.h, self.w-1)
        
        self.updateDims(screen)
        self.edges = ((0,self.w), (0, self.h - 1))
        
        self.reset(screen)
        
        self.start(screen)

    def reset(self, screen):
        self.start_time = 0
        self.game_over = False
        self.score = 1
        self.history = list()
        self.current_position = [self.w//2, self.h//2]
        if self.current_position[0]%2 != 0:
            self.current_position[0] += 1
        self.direction = "RIGHT"
        self.food_position = [0, 0]
        self.food_position = self.generate_food(True)
        screen.timeout(1)

    def init_logger(self):
        if not os.path.exists("snake_files"):
            os.makedirs("snake_files")
        
        self.logger = logging.getLogger("snake_logs")
        hdlr = logging.FileHandler("snake_files/snake_logs" + ".log")
        formatter = logging.Formatter('%(asctime)s %(levelname)s %(message)s')
        hdlr.setFormatter(formatter)
        self.logger.addHandler(hdlr)
        self.logger.setLevel(logging.DEBUG)
        
        self.logger.info("\n\n\nLogger initiated")

    def updateDims(self, screen):
        self.h, self.w = screen.getmaxyx()

    def start(self, screen):
        while True:
            while not self.game_over:
                self.update_window(screen)                    
            
            screen.timeout(-1)
            screen.clear()
            screen.addstr(0, 0, "Gameover")        
            screen.getch()
            self.reset(screen)

    def update_window(self, screen):
        current_time = time() - self.start_time
        
        thread = threading.Thread(target=self.generate_tail, args=(screen,))
        if current_time > 0.1:
            
            key = screen.getch()
            if key < 0:
                pass
            else:
                if key == cr.KEY_UP:
                    if self.direction != "DOWN":
                        self.direction = "UP"
                elif key == cr.KEY_DOWN:
                    if self.direction != "UP":
                        self.direction = "DOWN"
                elif key == cr.KEY_LEFT:
                    if self.direction != "RIGHT":    
                        self.direction = "LEFT"
                elif key == cr.KEY_RIGHT:
                    if self.direction != "LEFT":
                        self.direction = "RIGHT"
            
            self.history.insert(0, self.current_position.copy())
            self.history = self.history[:self.score]
            
            if self.direction == "UP":
                self.current_position[1] -= 1
            elif self.direction == "DOWN":
                self.current_position[1] += 1
            elif self.direction == "LEFT":
                self.current_position[0] -= 2
            elif self.direction == "RIGHT":
                self.current_position[0] += 2
            
            # Update screen
            screen.clear()
            food_pos = self.generate_food()
            try:
                screen.addstr(food_pos[1], food_pos[0], "FF")
                screen.addstr(self.current_position[1], self.current_position[0], "AA")
            except:
                # self.logger.info(f"Problem while adding string. Positions: {food_pos} , {self.current_position}")
                print("Error")
                
            thread.start()
            
            self.logger.info(f"{self.current_position} , {self.edges[1][1]}")
            
            if self.current_position[0] < self.edges[0][0] or self.current_position[0] > self.edges[0][1] or self.current_position[1] < self.edges[1][0] or self.current_position[1] > self.edges[1][1] or self.current_position in self.history:
                self.game_over = True
                return
                
            self.start_time = time()
            self.last_score = self.score
    
    def generate_tail(self, screen):
        for position in self.history[:self.score]:
            screen.addstr(position[1], position[0], "AA")
    
    def generate_food(self, ignore=False):
        
        if self.current_position == self.food_position or ignore:
            
            self.score += 1
            
            new_food_pos = None
            while True:
                new_food_pos = [random.randint(self.edges[0][0], self.edges[0][1]), random.randint(self.edges[1][0], self.edges[1][1])]
                if not new_food_pos == self.current_position and new_food_pos[0]%2==0:
                    break
            
            self.food_position = new_food_pos
                
            
            return new_food_pos
        else:
            return self.food_position

def main(screen):
    _ = Snake(screen)


if __name__ == '__main__':
    cr.wrapper(main)
